utils/treesitter/grammar.js

```diff
@@ -0,0 +1,507 @@
+          seq($.designator, '=', $._expression),
+          prec.left(seq($.where_clause, 'and', $.where_clause))
+        )
+      ),
+
+    where_expression: ($) =>
+      prec.left(PREC.TermPostfix, seq($._expression, 'where', $.where_clause)),
+
+    call_expression: ($) => prec(PREC.TermPostfix, seq($._expression, $.tuple)),
+
+    pointer_expression: ($) =>
+      prec(PREC.TypePostfix, seq($._expression, $.postfix_star)),
+
+    _expression: ($) =>
+      choice(
+        $.ident,
+        $.pointer_expression,
+        $.index_expression,
+        $.fn_type_expression,
+        $.designator,
+        seq('.', 'Self'),
+        $.literal,
+        $.builtin_type,
+        $.paren_expression,
+        $.array_literal,
+        $.if_expression,
+        $.binary_expression,
+        $.unary_prefix_expression,
+        $.as_expression,
+        $.paren_expression,
+        $.postfix_expression,
+        $.call_expression,
+        $.ref_expression,
+        $.deref_expression,
+        $.where_expression
+      ),
+
+    var_declaration: ($) =>
+      seq(
+        'var',
+        $._non_expression_pattern,
+        optional(seq('=', $._expression)),
+        ';'
+      ),
+
+    let_declaration: ($) =>
+      seq('let', $._non_expression_pattern, '=', $._expression, ';'),
+
+    assign_statement: ($) =>
+      seq($._expression, $._assign_operator, $._expression, ';'),
+
+    _assign_operator: ($) =>
+      choice('=', '+=', '/=', '*=', '%=', '-=', '&=', '|=', '^=', '<<=', '>>='),
+
+    match_clause: ($) =>
+      seq(choice(seq('case', $.pattern), 'default'), '=>', $.block),
+
+    match_statement: ($) =>
+      seq('match', '(', $._expression, ')', '{', repeat($.match_clause), '}'),
+
+    returned_var_statement: ($) => seq('returned', $.var_declaration, ';'),
+
+    while_statement: ($) => seq('while', '(', $._expression, ')', $.block),
+
+    break_statement: ($) => seq('break', ';'),
+
+    continue_statement: ($) => seq('continue', ';'),
+
+    return_statement: ($) =>
+      seq('return', optional(choice('var', $._expression)), ';'),
+
+    if_statement: ($) =>
+      seq('if', '(', $._expression, ')', $.block, optional($.else)),
+
+    else: ($) => choice(seq('else', $.if_statement), seq('else', $.block)),
+
+    for_statement: ($) =>
+      seq(
+        'for',
+        '(',
+        $._non_expression_pattern,
+        'in',
+        $._expression,
+        ')',
+        $.block
+      ),
+
+    statement: ($) =>
+      choice(
+        seq($._expression, ';'),
+        $.assign_statement,
+        $.var_declaration,
+        $.let_declaration,
+        $.match_statement,
+        $.returned_var_statement,
+        $.if_statement,
+        $.while_statement,
+        $.break_statement,
+        $.continue_statement,
+        $.return_statement,
+        $.for_statement
+      ),
+
+    _statement_list: ($) => seq($.statement, repeat($.statement)),
+
+    block: ($) => seq('{', optional($._statement_list), '}'),
```

Comment:
Simpler?

```suggestion
    _statement_list: ($) => repeat($.statement),

    block: ($) => seq('{', $._statement_list, '}'),
```

Reply:
rules are not allowed to match to empty string in treesitter. I think this is because of the error recovery mechanism.

Reply:
Then how about this:

```suggestion
    block: ($) => seq('{', repeat($.statement), '}'),
```